keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ResourceService.kt

```diff
@@ -0,0 +1,192 @@
+package com.netflix.spinnaker.keel.services
+
+import com.netflix.spinnaker.keel.api.DeliveryConfig
+import com.netflix.spinnaker.keel.api.Locatable
+import com.netflix.spinnaker.keel.api.Monikered
+import com.netflix.spinnaker.keel.api.Resource
+import com.netflix.spinnaker.keel.api.SimpleLocations
+import com.netflix.spinnaker.keel.api.SimpleRegionSpec
+import com.netflix.spinnaker.keel.api.artifacts.DeliveryArtifact
+import com.netflix.spinnaker.keel.api.id
+import com.netflix.spinnaker.keel.core.api.ResourceArtifactSummary
+import com.netflix.spinnaker.keel.core.api.ResourceSummary
+import com.netflix.spinnaker.keel.events.ApplicationActuationPaused
+import com.netflix.spinnaker.keel.events.ApplicationActuationResumed
+import com.netflix.spinnaker.keel.events.PersistentEvent
+import com.netflix.spinnaker.keel.events.ResourceActuationLaunched
+import com.netflix.spinnaker.keel.events.ResourceActuationPaused
+import com.netflix.spinnaker.keel.events.ResourceActuationResumed
+import com.netflix.spinnaker.keel.events.ResourceActuationVetoed
+import com.netflix.spinnaker.keel.events.ResourceCheckError
+import com.netflix.spinnaker.keel.events.ResourceCheckUnresolvable
+import com.netflix.spinnaker.keel.events.ResourceCreated
+import com.netflix.spinnaker.keel.events.ResourceDeltaDetected
+import com.netflix.spinnaker.keel.events.ResourceDeltaResolved
+import com.netflix.spinnaker.keel.events.ResourceMissing
+import com.netflix.spinnaker.keel.events.ResourceTaskFailed
+import com.netflix.spinnaker.keel.events.ResourceTaskSucceeded
+import com.netflix.spinnaker.keel.events.ResourceValid
+import com.netflix.spinnaker.keel.pause.ActuationPauser
+import com.netflix.spinnaker.keel.persistence.KeelRepository
+import com.netflix.spinnaker.keel.persistence.NoSuchDeliveryConfigException
+import com.netflix.spinnaker.keel.persistence.ResourceRepository.Companion.DEFAULT_MAX_EVENTS
+import com.netflix.spinnaker.keel.persistence.ResourceStatus
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.ACTUATING
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.CREATED
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.CURRENTLY_UNRESOLVABLE
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.DIFF
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.ERROR
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.HAPPY
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.PAUSED
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.RESUMED
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.UNHAPPY
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.UNKNOWN
+import com.netflix.spinnaker.keel.persistence.ResourceStatus.VETOED
+import org.springframework.stereotype.Component
+
+/**
+ * Service object that offers high-level APIs for resource-related operations.
+ */
```

Comment:
It seems like this class deals only with the resource history and status. I think that it makes sense to have a class that deals with those things - they have definitely outgrown living in the `ResourceRepository`. I'd suggest naming this class to reflect that it does those things - I don't think it's a catch-all for resource operations. Maybe `ResourceHistoryService` or `ResourceEventsService`?

I like separating these thing out into their own class because in the future we could decide to change how we do calculate status, and it would be easy to replace that implementation detail in this class.

Reply:
Yeah, it's hard to find the right boundaries for these things... I'm not sure having extremely fine-grained service objects (classes with one or two methods) is the best approach either, because you'll end up having to inject a lot of them in other components higher in the stack, like controllers, but I definitely agree that we don't want to create "catch-all" components -- that was my whole rationale for opposing `KeelRepository`.

Let me ask you a question back. You mentioned you liked `ApplicationService`, which started small but now has several more methods. My thinking here was to follow that pattern. What do you think is different in this case that warrants a more narrowly-scoped class versus that case?

Also curious if @robfletcher has any thoughts on this.

Reply:
I think the thing I like about `ApplicationService` is that it is build around a purpose (providing views for the application controller), vs just a collection of application-related methods. It has definitely grown to also include passthroughs for other methods, but the actual logic is all around rendering the views. I'd like to keep that same feeling here with the purpose of doing the event/history things. I don't like the grouping of "resource-related things" because it doesn't seem to have a focused purpose (and is more like the disliked `KeelRepository`).

I also like small and focused classes because they're easier to test and reason about. I don't mind injecting several to do a bigger thing. It's easy to refactor small classes.

Reply:
Fair enough. Let's go with `ResourceHistoryService`, then!